cluster-autoscaler/cloudprovider/kwok/kwok_provider.go

```diff
@@ -123,24 +125,28 @@ func (kwok *KwokCloudProvider) GetNodeGpuConfig(node *apiv1.Node) *cloudprovider
 
 // Refresh is called before every main loop and can be used to dynamically update cloud provider state.
 // In particular the list of node groups returned by NodeGroups can change as a result of CloudProvider.Refresh().
-// TODO(vadasambar): implement this
 func (kwok *KwokCloudProvider) Refresh() error {
 
-	// TODO(vadasambar): causes CA to not recognize kwok nodegroups
-	// needs better implementation
-	// nodeList, err := kwok.lister.List()
-	// if err != nil {
-	// 	return err
-	// }
+	allNodes, err := kwok.allNodesLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "failed to list all nodes from lister")
+		return err
+	}
+
+	targetSizeInCluster := make(map[string]int)
 
-	// ngs := []*NodeGroup{}
-	// for _, no := range nodeList {
-	// 	ng := parseAnnotationsToNodegroup(no)
-	// 	ng.kubeClient = kwok.kubeClient
-	// 	ngs = append(ngs, ng)
-	// }
+	for _, node := range allNodes {
+		ngName := getNGName(node, kwok.config)
```

Comment:
This will lead to `klog.Fatal` for cases where the node doesn't have the nodegroup label or annotation. 

You might have to use a filter function to filter only nodes which have the nodegroup label/annotation OR convert `klog.Fatal` into an non-fatal error log.

Reply:
ok. I converted `klog.Fatal` to `klog.Warning` and added comments to `getNGName`.